docs/rfcs/20221018_cluster_bootstrap.md

```diff
@@ -0,0 +1,325 @@
+
+Alternate approaches to controller group initialization
+
+
+
+* Rather than waiting for an explicit init command, unconditionally proceed once we know all seed servers are healthy and configured uniformly.
+    * **Upside:** deployment is easier to automate because there is no special step to run init.
+    * **Downside:** we no longer have an explicit step with which to provide initial cluster-wide configs.
+    * **Downside:** this is still error prone in the case of all seed servers being wiped and restarted, e.g. the seed servers are haphazardly placed in the same region and there is a region failure. While more difficult, it’s still possible to restart some nodes empty and start up with two clusters (compared to the current proposal, if the seeds start empty, they would still need to await the init command, rather than start a new cluster).
+* Have the default for `wait_for_init` to be false
+    * **Downside:** as a default, this would be unsafe for the reasons stated above.
+
+Alternate approaches to `initial_cluster_info` endpoint
+
+
+
+* Use an entirely separate server bound to a different port
+    * **Upside:** likely better tested approach
+    * **Downside:** may require additional configuration
+* Use configs for the RPC server, but once we’re done, stop the initial_cluster_info server, and then start a new server with the runtime protocol
+    * **Downside:** there may be a brief blip where no RPC server is unavailable
+    * **Downside:** unclear whether this approach is any better tested
+
+
+### **Single-node cluster formation**
+
+It doesn’t seem unreasonable that some clusters knowingly start out with cluster size 1, and then grow as use cases grow more mature on the cluster. Additionally, the existing user experience for developer clusters relies heavily on being able to not provide a seed server list.
+
+To that end, we will continue to support supplying an empty seed servers list or a seed server list of size one, which will be construed to mean starting a cluster with size 1 node.
+
+Alternatives
+
+
+
+* Don’t support single node clusters period.
+    * **Downsides:** use case doesn’t sound unreasonable, especially for folks just testing out a binary. A goal of this project is to make trying out Redpanda easy and failsafe, and single-node/local clusters seem like an important part of that experience.
+* Force users to specify when they want a single node cluster by setting some `minimum_initial_cluster_size` configuration
+
+
+### **Controller log initialization**
+
+A new controller log command will be added to indicate the initial desired state of the cluster, including initial admin secrets, license, etc. This will be a new controller command with roughly the following payload:
+
+
+```
+struct cluster_init_cmd_data {
+    string cluster_uuid;
+    vector<tuple<uuid, node_id, addr>> seed_servers;
```

Comment:
This one didn't make it. Instead, the initial `raft::configuration` that precedes this command is the one that has this.